piker/brokers/kucoin.py

```diff
@@ -0,0 +1,718 @@
+        )
+
+
+@trio_async_generator
+async def stream_messages(
+    ws: NoBsWs, sym: str
+) -> AsyncGenerator[NoBsWs, dict]:
+    timeouts = 0
+    last_trade_ts = 0
+
+    while True:
+        with trio.move_on_after(3) as cs:
+            msg = await ws.recv_msg()
+        if cs.cancelled_caught:
+            timeouts += 1
+            if timeouts > 2:
+                log.error('kucoin feed is sh**ing the bed... rebooting...')
+                await ws._connect()
+
+            continue
+
+        if msg.get('subject'):
+            msg = KucoinMsg(**msg)
+            match msg.subject:
+                case 'trade.ticker':
+                    trade_data = KucoinTrade(**msg.data)
+
+                    # XXX: Filter for duplicate messages as ws feed will
+                    # send duplicate market state
+                    # https://docs.kucoin.com/#level2-5-best-ask-bid-orders
+                    if trade_data.time == last_trade_ts:
+                        continue
+
+                    last_trade_ts = trade_data.time
+
+                    yield 'trade', {
+                        'symbol': sym,
+                        'last': trade_data.price,
+                        'brokerd_ts': last_trade_ts,
+                        'ticks': [
+                            {
+                                'type': 'trade',
+                                'price': float(trade_data.price),
+                                'size': float(trade_data.size),
+                                'broker_ts': last_trade_ts,
+                            }
+                        ],
+                    }
+
+                case 'level2':
+                    l2_data = KucoinL2(**msg.data)
+                    first_ask = l2_data.asks[0]
+                    first_bid = l2_data.bids[0]
+                    yield 'l1', {
+                        'symbol': sym,
+                        'ticks': [
+                            {
+                                'type': 'bid',
+                                'price': float(first_bid[0]),
+                                'size': float(first_bid[1]),
+                            },
+                            {
+                                'type': 'bsize',
+                                'price': float(first_bid[0]),
+                                'size': float(first_bid[1]),
+                            },
+                            {
+                                'type': 'ask',
+                                'price': float(first_ask[0]),
+                                'size': float(first_ask[1]),
+                            },
+                            {
+                                'type': 'asize',
+                                'price': float(first_ask[0]),
+                                'size': float(first_ask[1]),
+                            },
+                        ],
+                    }
+
+
+@acm
+async def open_history_client(
+    symbol: str,
+    type: str = '1m',
```

Comment:
sorry just noticed this..
try not to use built-ins as var names if you can, also i don't think this is used in the func body?

Reply:
yeah i think you can just drop it no?

`timeframe` info is passed into the delivered closure on call.